### PR TITLE
[@types/ws] Update ws definitions to v8.2

### PR DESCRIPTION
**Pull request by tamuratak**
Update `ws` definitions to v8.2:

- https://github.com/websockets/ws/releases/tag/8.0.0
- https://github.com/websockets/ws/releases/tag/8.1.0
- https://github.com/websockets/ws/releases/tag/8.2.0
- Add the RawData type.
- Rename the OpenEvent class to Event.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/websockets/ws/releases/tag/8.0.0
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by Pokute** [comment]
These are not correct TS definitions for ES Module.
The ES module has completely different exports, see https://github.com/websockets/ws/blob/78adf5f7737b10166c6f4903e6a5cdbbbb59934c/wrapper.mjs

**Comment by tamuratak** [comment]
When a CommonJS main file and an ES module main file have different types, we don't have much we can do. The current PR can work with `wrapper.mjs` except for some cases.

@Pokute do you have any suggestions to improve?

**Comment by Pokute** [comment]
I couldn't access `WebSocketServer `through the definition file in this PR since the commonJS main file you linked with `WebSocket.Server = require('./lib/websocket-server');` -line is not when imported as an ES module.

I did write a [version of the definition, based on this version, that works for ES modules](https://gist.github.com/Pokute/5486d8a326e1a2579425df60dd7e1643) and also has typings for Receiver and Sender. I didn't test them at all though.

**Comment by tamuratak** [comment]
>  I didn't test them at all though. Feel free to use it as you see fit if it helps.

Thank you, but I won't use. I don't like to make breaking changes in this PR.

>  I do not have any idea how to have different definition files for ESM and CJS, though.

Since there is no way to improve this PR, I want to wait for the maintainers' review without changes.

**Comment by tamuratak** [comment]
The core committer of `ws` says

> https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55151#issuecomment-899149056  makes sense to me now that ws uses an ES module wrapper. It does not seem bad to me to cut a new major version of @types/ws that only works with ws@>=8. This is why semver versioning exists.

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:176]: Is it possible to add `WebSocketServer` as alias for `Server` inside this namespace? I think something like this

```js
class WebSocketServer extends Server {}
```

should work. I'm not sure if there are better options, and ofc it will only work with `ws@>=8.2.0`.

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:284]: This is technically a breaking change as `protocols` is an array in `ws@<8`.

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:106]: In `ws@8` this is a noop. The `handleProcotols` type change is already a breaking change so I wonder if it makes sense to remove this.

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:118]:
```suggestion
    on(event: "message", listener: (this: WebSocket, data: WebSocket.RawData, isBinary: boolean) => void): this;
```

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:142]:
```suggestion
    off(event: "message", listener: (this: WebSocket, data: WebSocket.RawData, isBinary: boolean) => void): this;
```

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:154]:
```suggestion
    addListener(event: "message", listener: (data: WebSocket.RawData, isBinary: boolean) => void): this;
```

File: types/ws/index.d.ts
Comment (lpinca) [inline types/ws/index.d.ts:166]:
```suggestion
    removeListener(event: "message", listener: (data: WebSocket.RawData, isBinary: boolean) => void): this;
```

**Review by lpinca — COMMENTED** [review]
The reason is the same as https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55151/files#r690233908. The are already breaking changes so I think it is better to be more strict.

**Comment by vaporox** [comment]
I'd appreciate if you could add the [named exports](https://github.com/websockets/ws/blob/master/wrapper.mjs#L7) as well, currently there's only one `export =` in the declaration file and that can't be imported without enabling [`allowSyntheticDefaultImports`](https://www.typescriptlang.org/tsconfig#allowSyntheticDefaultImports).

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts]: Removed.

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts:283]: A `@types` definition file should be used with the same `major.minor` version. This PR will be used only with `ws@8`. The change would not be an issue.

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts:174]: I will add the line after `ws@8.2.0` published.

**Comment by tamuratak** [comment]
> I'd appreciate if you could add the named exports as well, currently there's only one export = in the declaration file and that can't be imported without enabling allowSyntheticDefaultImports. They also exist in the main file, so it's not in the ESM wrapper only.

That is out of the cope of this PR. Please open a new PR.

File: types/ws/index.d.ts
Comment (lpinca) [reply types/ws/index.d.ts:174]: `ws@8.2.0` is out. `WebSocket.WebSocket` should also be added.

Comment (tamuratak) [reply types/ws/index.d.ts:176]: Thanks.

**Review by lpinca — APPROVED** [review]

File: types/ws/index.d.ts
Comment (rbuckton) [inline types/ws/index.d.ts:174]: You can do this using TS namespace imports:
```suggestion
import WebSocketAlias = WebSocket;
```
And below,
```ts
export WebSocket = WebSocketAlias;
```

File: types/ws/index.d.ts
Comment (rbuckton) [inline types/ws/index.d.ts:350]:
```suggestion
    export WebSocket = WebSocketAlias;
```

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts:350]: Your suggestion doesn't work well.

```diff
diff --git a/types/ws/index.d.ts b/types/ws/index.d.ts
index 876496d06f..3760cf3d91 100644
--- a/types/ws/index.d.ts
+++ b/types/ws/index.d.ts
@@ -171,7 +171,7 @@ declare class WebSocket extends EventEmitter {
     removeListener(event: string | symbol, listener: (...args: any[]) => void): this;
 }
 
-declare class WebSocketAlias extends WebSocket {}
+import WebSocketAlias = WebSocket;
 
 declare namespace WebSocket {
     /**
@@ -347,7 +347,7 @@ declare namespace WebSocket {
     }
 
     class WebSocketServer extends Server {}
-    class WebSocket extends WebSocketAlias {}
+    export type WebSocket = WebSocketAlias;
 
     // WebSocket stream
     function createWebSocketStream(websocket: WebSocket, options?: DuplexOptions): Duplex;
```

```
$  npm test ws

> definitely-typed@0.0.3 test /Users/tamura/src/github/DefinitelyTyped
> dtslint types "ws"

Error: /Users/tamura/src/github/DefinitelyTyped/types/ws/index.d.ts:350:5
ERROR: 350:5  strict-export-declare-modifiers  'export' keyword is redundant here because all declarations in this module are exported automatically. If you have a good reason to export some declarations and not others, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

    at /Users/tamura/src/github/DefinitelyTyped/node_modules/dtslint/bin/index.js:207:19
    at Generator.next (<anonymous>)
    at fulfilled (/Users/tamura/src/github/DefinitelyTyped/node_modules/dtslint/bin/index.js:6:58)
npm ERR! Test failed.  See above for more details.
```

Comment (tamuratak) [reply types/ws/index.d.ts:350]: Your suggestion doesn't work well.

VS Code says the following line is invalid:
```ts
export WebSocket = WebSocketAlias;
```

Comment (tamuratak) [reply types/ws/index.d.ts:350]: @rbuckton I had better make changes as suggested even if VS Code says it is invalid?

**Comment by onury** [comment]
@tamuratak any chance for a release soon? 
Thanks for your work.

File: types/ws/index.d.ts
Comment (onury) [reply types/ws/index.d.ts:350]: @rbuckton is there any down sides to @tamuratak's version?

Comment (tamuratak) [reply types/ws/index.d.ts:350]: @rbuckton ping

**Comment by beenotung** [comment]
The fix is working for me (just upgrading to ws^8.2.2)

**Comment by beenotung** [comment]
In case anyone is waiting to use ws@8 from typescript before this pull request is approved by DefinitelyTyped, I published a typestub package on npm: https://www.npmjs.com/package/typestub-ws

**Comment by talentlessguy** [comment]
what's blocking this PR to be merged?

File: types/ws/index.d.ts
Comment (vansergen) [reply types/ws/index.d.ts:174]: Or

```suggestion
declare const WebSocketAlias: typeof WebSocket;
```

File: types/ws/index.d.ts
Comment (vansergen) [inline types/ws/index.d.ts:349]:
```suggestion
    const WebSocketServer: typeof Server;
```

**Comment by tamuratak** [comment]
@loyd @mlamp @TitaneBoy @reduckted @teidesu @wojtkowiak @k-yle 

Can anyone else can review this PR? @rbuckton does not respond.

File: types/ws/index.d.ts
Comment (k-yle) [inline types/ws/index.d.ts:350]: @tamuratak This seems to work well:

```suggestion
    type WebSocketServer = Server;
    type WebSocket = WebSocketAlias;
```

Are there any other issues blocking this PR from being merged?

**Review by k-yle — CHANGES_REQUESTED** [review]

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts:350]: Thank your. It doesn't. I have added the following tests:

```ts
{
    const ws: WebSocket.WebSocket = new WebSocket.WebSocket("ws://www.host.com/path");
}

{
    const wss: WebSocket.WebSocketServer = new WebSocket.WebSocketServer({ port: 8081 });
}
```

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts:349]: It doesn't work in the following test:

```ts
{
    const wss: WebSocket.WebSocketServer = new WebSocket.WebSocketServer({ port: 8081 });
}
```

File: types/ws/index.d.ts
Comment (tamuratak) [reply types/ws/index.d.ts:350]: > Are there any other issues blocking this PR from being merged?

No, there aren't.

Comment (k-yle) [reply types/ws/index.d.ts:350]: I see, that's unfortunate. Those tests seem to pass with [this suggestion](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55151#discussion_r693430070), so that change could be applied.

Is there actually an issue with the way these two lines are currently defined? If not, I see no point in trying to find a more idiomatic solution to a problem that has already been solved.

Comment (tamuratak) [reply types/ws/index.d.ts:350]: I think there is no real issue with the current pull request so far. We had better go forward with merging the pull request.

Comment (k-yle) [reply types/ws/index.d.ts:350]: :+1:, Unless someone can point out an issue with this approach then this should be good to merge (cc @rbuckton)

**Review by k-yle — APPROVED** [review]

**Review by lpinca — APPROVED** [review]

File: types/ws/index.d.ts
Comment (weswigham) [reply types/ws/index.d.ts:350]: @rbuckton's code is missing a `const` or the like, eg
```ts
export const WebSocket = WebSocketAlias;
```
but that won't also reexport the types - you'd need
```ts
export {WebSocketAlias as WebSocket};
```
for that, ideally, but that's not allowed within a namespace declaration, so the best you could do is probably
```ts
export const WebSocket = WebSocketAlias;
export type WebSocket = WebSocketAlias;
```
to export both meanings. It's probably better than the class redeclarations that imply a prototype hierarchy that doesn't exist.

**Review by weswigham — CHANGES_REQUESTED** [review]

**Comment by tamuratak** [comment]
> so the best you could do is probably

```ts
export const WebSocket = WebSocketAlias;
export type WebSocket = WebSocketAlias;
```

> to export both meanings. It's probably better than the class redeclarations that imply a prototype hierarchy that doesn't exist.

It doesn't work.


スクリーンショット 2021-09-25 10 46 04


スクリーンショット 2021-09-25 10 46 16

**Comment by weswigham** [comment]
Geh, sorry, `export const: WebSocketAlias;` - I listed the syntax for a non-declaration file.

**Comment by tamuratak** [comment]
> Geh, sorry, export const: WebSocketAlias; for the const line - I listed the syntax for a non-declaration file.

It doesn't work.

スクリーンショット 2021-09-25 18 00 49

**Comment by weswigham** [comment]
Omg the train of typos - `export const WebSocket: WebSocketAlias;`

**Comment by tamuratak** [comment]
> Omg the train of typos - export const WebSocket: WebSocketAlias;

It doesn't work.

スクリーンショット 2021-09-25 18 23 42

スクリーンショット 2021-09-25 18 24 22

**Comment by weswigham** [comment]
:facepalm: `typeof WebSocketAlias`

**Comment by tamuratak** [comment]
I see.

File: types/ws/index.d.ts
Comment (vansergen) [inline types/ws/index.d.ts:350]:
```suggestion
    const WebSocketServer: typeof Server;
```

**Review by vansergen — APPROVED** [review]

**Review by weswigham — APPROVED** [review]

**Comment by tamuratak** [comment]
Ready to merge

**Pull request merged**
